File: sdk/dotnet/NotchAgent.Desk.Protocol.Tests/DeskFrameCodecTests.cs
using System.Text;
using NotchAgent.Desk.Protocol;

namespace NotchAgent.Desk.Protocol.Tests;

public sealed class DeskFrameCodecTests
{
    [Fact]
    public void FrameRoundTrips()
    {
        var payload = Encoding.UTF8.GetBytes("{\"product\":\"NotchAgent Desk\",\"nonce\":42}");
        var encoded = DeskFrameCodec.Encode(new DeskFrame(DeskFrameType.Hello, 7, payload));

        Assert.Equal(0, encoded[^1]);
        var decoded = DeskFrameCodec.DecodePacket(encoded.AsSpan(0, encoded.Length - 1));
        Assert.Equal(DeskFrameType.Hello, decoded.Type);
        Assert.Equal(7u, decoded.Sequence);
        Assert.Equal(payload, decoded.Payload);
    }

    [Fact]
    public void PartialStreamRoundTrips()
    {
        var encoded = DeskFrameCodec.Encode(new DeskFrame(DeskFrameType.Heartbeat, 9, [0, 1, 0, 2]));
        var decoder = new DeskFrameStreamDecoder();

        Assert.Empty(decoder.Append(encoded.AsSpan(0, 3)));
        var frames = decoder.Append(encoded.AsSpan(3));
        Assert.Single(frames);
        Assert.Equal(DeskFrameType.Heartbeat, frames[0].Type);
    }

    [Fact]
    public void CorruptedPacketIsRejected()
    {
        var encoded = DeskFrameCodec.Encode(new DeskFrame(DeskFrameType.Snapshot, 1, Encoding.UTF8.GetBytes("safe")));
        encoded[4] ^= 0x20;
        Assert.Throws<DeskProtocolException>(() => DeskFrameCodec.DecodePacket(encoded.AsSpan(0, encoded.Length - 1)));
    }
}

File: sdk/apple/Tests/NotchAgentDeskProtocolTests/DeskFrameCodecTests.swift
import Foundation
import Testing
@testable import NotchAgentDeskProtocol

@Test func frameRoundTrip() throws {
    let payload = Data(#"{"product":"NotchAgent Desk","nonce":42}"#.utf8)
    let encoded = try DeskFrameCodec.encode(.init(type: .hello, sequence: 7, payload: payload))
    #expect(encoded.last == 0)
    let decoded = try DeskFrameCodec.decodePacket(encoded.dropLast())
    #expect(decoded == .init(type: .hello, sequence: 7, payload: payload))
}

@Test func partialStreamRoundTrip() throws {
    let encoded = try DeskFrameCodec.encode(.init(type: .heartbeat, sequence: 9, payload: Data([0, 1, 0, 2])))
    var decoder = DeskFrameStreamDecoder()
    #expect(decoder.append(encoded.prefix(3)).isEmpty)
    let result = decoder.append(encoded.dropFirst(3))
    #expect(result.count == 1)
    #expect(try result[0].get().type == .heartbeat)
}

@Test func checksumFailureIsRejected() throws {
    var encoded = try DeskFrameCodec.encode(.init(type: .snapshot, sequence: 1, payload: Data("safe".utf8)))
    encoded[encoded.index(encoded.startIndex, offsetBy: 4)] ^= 0x20
    #expect(throws: (any Error).self) {
        try DeskFrameCodec.decodePacket(encoded.dropLast())
    }
}

